### ISSUE TITLE
Performance of Http2Stream.setProperty(...) / .getProperty(...).

### ISSUE THREAD

**Issue by buchgr** [netty/netty#3570]
I have been doing some benchmarking and profiling on the HTTP/2 codec. One of the pain points is that the codec is spending too much time and memory on `Http2Stream.setProperty(...) / .getProperty(...)`. In my benchmarks we spend roughly 11% of the CPU time on retrieving and putting "properties" and use just over 22% of the heap for `HashMaps`.

The main reason that we spend so much time on the properties is because of the local and remote `FlowState`. We look it up all the time. The main reason why we are using so much memory for `HashMap` is that each stream uses its own.

My preferred way to solve this is:

* Add `remoteFlowState()` and `localFlowState()` properties to `Http2Stream`. Use those instead of the properties.
* Make `(set|get)Property` package private and initialize the `HashMap` backing this lazily (right now that's only used in `CompressorHttp2ConnectionEncoder` and `DelegatingDecompressorFrameListener`). That way we only pay the overhead of a `HashMap` per stream when those classes are used and we have the flexibility to attach things to the stream in future code.
* Introduce public interface methods `setUserProperty(Object val) / getUserProperty()`. Those are not used from within the codec package, but a user is free to attach whatever object he feels like.

WDYT @nmittler @Scottmitch

**Comment by nmittler** [comment]
@buchgr good find ... yeah, @Scottmitch and I sort of punted on this in favor of a flexible API :)

I'll just recap my idea that we discussed earlier as a way to keep the flexibility but not sacrifice performance:

Instead of a map, each `Stream` could maintain generic properties as an array, indexed by a unique property key. This key could be obtained via the connection:

```java
interface Http2Connection {
...
   // The first one is zero and increments each time it's called.  It's just the index for the next property.
    int nextPropertyKey();
...
}
``` 

The stream interface could then be changed to something like this:

```java
interface Http2Stream {
    // Throws if the key is above the last valid key returned by the connection. Grows the array
    // as necessary.
    void setProperty(int key, Object value);

    // Throws if the key is invalid or if the value is not of type clazz.
    <T> T getProperty(int key, Class<T> clazz);
}
```

Internally, the stream could just use a plain array to store the properties.  It might default to null and only be allocated as necessary. For the first allocation, it probably makes sense to use a reasonable size, like 4 or 5.

WDYT?

**Comment by buchgr** [comment]
@nmittler 

> I'll just recap my idea that we discussed earlier as a way to keep the flexibility but not sacrifice performance:

With my suggestion we would maintain the same flexibility (see point 2, it's different than what we discussed). 

Also I can see two things that will likely have a performance impact (Don't know how much though):
 1.  `(local|remote)FlowState()` and `(set|get)UserProperty()` can and will be inlined easily. The array approach not so much.
 2. One does pay a (smaller) memory overhead because the array will likely be not sized ideally and if it is, any addition of a stream will lead to copying the array. Yes those arrays are small, but if you have a lot of short-lived streams (like RPCs) they do account for something.

Also I think the `nextPropertyKey()` method should be on the `Http2Stream`? Not every stream will want to use the same property / the same number of properties. So that would introduce unnecessary gaps in the backing arrays?

My main concern with the approach though is that the API gets very complex to use. 
Also how would you make sure that someone doesn't overwrite the property key of someone else e.g. a user does `stream.setProperty(0, new UserObject())` and overwrites one of the `FlowState`s. Now we could enforce that a property can only be written once, but then that property could never be updated or nulled in case the stream lives on but that attached object should be garbage collected.

**Comment by Scottmitch** [comment]
@buchgr
+1 on this find! I think we can optimize/cleanup up here :)
+1 on making `remoteFlowState()` and `localFlowState()` directly accessible.  These are part of the codec...not sure if it makes sense to be accessible outside the codec but we should have quick access to them.  The read only members maybe fine but we should be careful about if we are exposing the write interfaces...

I wonder if we can have a more flexible stream creation mechanism to account for extensibility.  If so we may not even need to support the additional features as we previously discussed @nmittler.  For example if the `createStream(...)` could be more easily intercepted and then decorated we only have to support what we need where we need it, and the user can use this same mechanism.  No maps, or arrays...

```java
public interface Http2ModifiableStream {
  // all of the methods that DefaultHttp2Connection.DefaultStream has that we depend upon
}

public interface Http2ModifiableStreamFactory {
  void parent(Http2ModifiableStreamFactory parent);
  Http2ModifiableStreamFactory parent();

  Http2ModifiableStream newInstance(streamId, state);
  Http2ModifiableStream wrap(Http2ModifiableStream stream);
}

public abstract class Http2ModifiableStreamFactoryBase {
   private Http2ModifiableStreamFactory parent;

  void parent(Http2ModifiableStreamFactory parent) {
    this.parent = parent;
  }

  Http2ModifiableStreamFactory parent() {
    return parent;
  }

  Http2ModifiableStream wrap(Http2ModifiableStream stream) {  
    for (Http2ModifiableStreamFactory curr = parent;curr != null; curr = curr.parent) {
      stream = curr.wrap(stream);
   }
   return stream;
}
```

In the DefaultHttp2Connection.DefaultEndpoint (where we are doing `new DefaultStream(..)`) we could use this factory.

```java
// On the Endpoint or connection we would want the following method...
private Http2ModifiableStreamFactory streamFactory;
void streamFactory(Http2ModifiableStreamFactory factory) {
  if (streamFactory == null) {
    this.streamFactory = factory;
  } else {
    this.streamFactory.parent(factory);
  }
}


....
// instead of new DefaultStream()
Http2ModifiableStream stream = factory.newInstance(streameId, state);
```

**Comment by Scottmitch** [comment]
@nmittler @buchgr - I updated the above code....the idea is pretty "fresh" right now...so it is just to demonstrate an idea of allowing decoration to keep our "core" interfaces small...and let the user decorate/extend to provide other methods they need....the compressor/decompressor could use this too.

**Comment by Scottmitch** [comment]
It would be much simpler just to allow the user to extend `DefaultStream` to some degree (and use a more typical factory pattern)...because when the user (or anyone relying on this mechanism) wants to get the "decorator" object they may have to go through the "parent" links to get the right type.

**Comment by buchgr** [comment]
@Scottmitch 

I think the "chain of decorators" suggestion is quite complex and cumbersome. It would also make accessing the property bascially a linked list iteration.

 Allowing someone to provide a `Http2StreamFactory` to the connection constructor and making `DefaultStream` public is a possibility. However, that would also mean that we need a different `Http2Stream` class for whenever we want to "attach" a different property. So we would already need one additional stream class in the codec (`CompressionHttp2Stream` or somethin). If there is another part of the codec that needs a property then we would need yet another class `NewPropertyHttp2Stream`. If someone wants to use both parts and thus needs access to both properties we need `CompressionNewPropertyHttp2Stream`. Also since the `DelegatingCompressorFrameListener` is a ... well ... listener, we can't ensure that someone always passes in a `CompressionHttp2StreamFactory` to the connection constructor. 

Could you provide feedback on @nmittler's and my suggestion please?

**Comment by Scottmitch** [comment]
@buchgr - Good points. I would like to avoid the array/map interface entirely. It is flexible but I don't think the overhead is justified...I would favor an object that the user can manually control (and make it a map if they want). Let me recall what we had at one point to rationalize where we are now.... (@nmittler correct me if I misremember):

```java
public Http2Stream {
    // User extension points
    <T> void userProperty(T obj);
    <T> T userProperty();

    // Flow control stuff...interfaces may have been the same at one point
    Http2RemoteFlowState remoteState();
    void remoteState(Http2RemoteFlowState state);
    Http2LocalFlowState localState();
   void localState(Http2LocalFlowState state);

    // For encoder/decoder
    void decompressor(EmbeddedChannel channel);
    EmbeddedChannel decompressor();
    void compressor(EmbeddedChannel channel);
    EmbeddedChannel compressor();
}
```

We moved away from this I think because:
1. FlowController interfaces were in flux...and we didn't really want to expose them externally.  We have a common base class at one point...and maybe even moved a `FlowState flowState();` into the `Endpoint` class.
2. The encoder/decoder is not really core to the codec and I just stuck it there for ease of access.
3. The user object stuff was superseded after we looked at all of these issues and decided just to provide a flexible map like interface that the users could also use.

Here is some mitigation options:
1. Derive read only interfaces for the flow control states (just like we do for Http2Stream) and expose getters/setters on the connection (or the endpoint if that makes sense).  Create a concrete class and set them on the connection.  So we don't have to worry about them being exposed...just like we don't worry about `Http2Stream` being exposed. (this is in agreement with @buchgr's first bullet point).
2.  Hum.  A few options:
2.1 Make the `DefaultStream` package private.  Then we modify this object to suite the needs of the codec.  This is nice because it is not exposed to the user via `Http2Stream`.
2.2 We give the user an optional property....why don't we give the codec an optional property?
```java
public class Http2Stream {
  <T> codecProperty(T obj);
  <T> T codecProperty();
}
```
Then the encoder/decoder could store an array of size 2 (who ever finds it as null could initialize it). Then choose a convention in the coded as to how this is accessed (i.e. Encoder uses index `0` and Decoder uses `1`).  If there are more elements of the codec that want to take advantage of this in the future...then we will have to ensure they can all play nice together.
3. Lets go back to the `<T> void userProperty(T obj);` methods.  The user can put an array/map in here if they want. (this is in agreement with @buchgr's last bullet point).

**Comment by buchgr** [comment]
Thanks @Scottmitch!  

I agree with you in points 1 and 3. Let's see what @nmittler thinks about it. (I will implement those two points quickly and see how it affects performance.).

> 2.1 Make the DefaultStream package private. Then we modify this object to suite the needs of the codec. This is nice because it is not exposed to the user via Http2Stream.

This seems like the best option :sob:

**Comment by Scottmitch** [comment]
@nmittler +1.  I think 2.1 is the lesser of the evals so far.

**Comment by nmittler** [comment]
@Scottmitch good discussion!

As it stands, a codec configured for compression would require 2 additional properties of its own (one for compression and one for decompression).  So I'm not sure that `T codecProperty()` would quite cut it here.  I think we need something a little more flexible.

In general, I'm fine with the flow control states being top-level properties of `Http2Stream`. I'm also fine with having a `getUserProperty`.  The thing that bugs me is how to cleanly support the expansion of both codec and user-defined properties.

I'm liking the hybrid approach that was mentioned earlier, where flow states are top-level properties and then we just have a lazy-created array for all other properties (user and codec).  The array can simply be based on the highest reserved key in the connection. When a stream is created, it can optionally allocate the array based on this size ... in this way it is unlikely that any array would ever have to be resized and there would be no wasted indices.

**Comment by Scottmitch** [comment]
> As it stands, a codec configured for compression would require 2 additional properties of its own (one for compression and one for decompression)

Agreed. The suggestion in https://github.com/netty/netty/issues/3570#issuecomment-89092942 was to use an array that the codec would manage to account for this:

> store an array of size 2 (who ever finds it as null could initialize it). Then choose a convention in the coded as to how this is accessed (i.e. Encoder uses index 0 and Decoder uses 1).

> I'm liking the hybrid approach that was mentioned earlier, where flow states are top-level properties and then we just have a lazy-created array for all other properties (user and codec).

Noted.  I'm still of the camp that things in the codec (in the `io.netty.handler.codec.http2` package) could do things internally via DefaultStream (or something like it...), and then users external to the codec can maintain their own property object.  This way we have less coordination and coupling with how the users attach things to a stream, and how the codec internally uses the stream.

**Comment by nmittler** [comment]
> Noted. I'm still of the mindset that things in the codec (in the io.netty.handler.codec.http2 package) could do things internally via DefaultStream (or something like it...), and then users external to the codec can maintain their own property object. This way we have less coordination and coupling with how the users attach things to a stream, and how the codec internally uses the stream.

And as the codec grows and requires new properties? Do we just keep adding fields to `DefaultStream`? ... I'm not sure that's a good path for us.  The same applies for the user property ... maybe they need multiple.  I guess I'm not quite sure what is causing you heartburn about that idea? It's not wasting space and lookup time will be O(1).

**Comment by Scottmitch** [comment]
> And as the codec grows and requires new properties? Do we just keep adding fields to DefaultStream? 

If it is internal to codec I don't think that is the worst thing. My mental model for these types of changes are internal to the package. I agree not as flexible in terms of API, but I'm not sure if it is worth putting stuff in the public API that the users also have to worry about coordinating with the codec (sizes of arrays, types of object is which indices, etc...).

> The same applies for the user property ... maybe they need multiple.

I guess I'm envisioning the user property as *the* extension point for users.  If we provide an interface as an object...they can use w/e type they want to manage it.  Collection, Map, ConcurrentMap, FooObject.  This way the extensibility is up to them...and it is easy to keep this consistent when changes occur because we don't depend upon it.

> I guess I'm not quite sure what is causing you heartburn about that idea?

Perhaps if I see it maybe I will like it :)  What will the type of the array be?  Will this require "unsafe" casting every time it is accessed?  How will the size be configured from the user point of view...and how will this interact with the codec pieces to get the size correct and avoid resize/copy?

**Comment by buchgr** [comment]
@Scottmitch @nmittler 

FYI with the flowstate and the userproperty directly attached to the object I am seeing a consistent 6% increase in throughput, 50percentile latency is down also 6%, 90percentile latency is down 7%. `HashMap` disappears from the hot allocations.

**Comment by Scottmitch** [comment]
@buchgr sounds good...just to confirm from https://github.com/netty/netty/issues/3570#issuecomment-89095795 this test scenario excludes code for the compressor and decompressor?

**Comment by buchgr** [comment]
@Scottmitch yes. I left the `(set|get)Property` untouched and just didn't use them.

@nmittler 
> The same applies for the user property ... maybe they need multiple.

As @Scottmitch said a user is free to plug in whatever, be it his own "property object" or a map.

> I guess I'm not quite sure what is causing you heartburn about that idea? It's not wasting space and lookup time will be O(1).

I think the API is just way too complex with having to request property keys and such. In terms of performance, I don't know. While fields and arrays are both constant time access, my guess would be that that constant for fields is a lot smaller, but I don't know. Would have to benchmark.

Also something to consider is that we probably have to make that user object field volatile, and there is no way to make an array element volatile in Java. Anyways, my main concern is really the API and I don't see any downside with the user object. Having that user object just "feels natural" to me and not like a hack.

@nmittler @Scottmitch 

Regarding what to do about "codec properties". While I am fine with just adding package private setters / getters, I would also be fine with just keeping the current hashmap for that. Just make `(set|getProperty)` package private. Yes, it's a bit slower. Yes, it uses more memory. However, it's not used much and as long as nobody complains we can likely get away with it.

Regarding the flow state. I think we all agree on having it as a field.

Regarding the user property. I would argue, similar to the flow state, every application using the HTTP/2 codec will make use of it. One will just always need to maintain some state / information per stream, because the way the HTTP/2 codec is written, there are no channelhandlers. In a "natural Netty codec" every channelhandler has a few fields where it maintains state, in a HTTP/2 application one would have to attach it to the stream instead. So that's why I feel it also makes sense to have the user object as a field.

**Comment by nmittler** [comment]
@Scottmitch @buchgr I have some ideas for how we might be able to keep the API clean and still be generic.  I'll try to put together a PR today.

**Comment by nmittler** [comment]
@Scottmitch @buchgr I've put together a sketch of my idea in #3593

**Comment by Scottmitch** [comment]
@nmittler - Thanks! I left a comment there for discussion.

**Comment by Scottmitch** [comment]
@nmittler - Assigned to you bcz you have been actively working in this area with https://github.com/netty/netty/pull/3593 ...but @buchgr said in https://github.com/netty/netty/pull/3593#issuecomment-90727726 that he is working on a PR to resolve this issue.  Feel free to re-assign to me.

**Comment by buchgr** [comment]
@Scottmitch yeah. will submit tomorrow!

**Comment by nmittler** [comment]
@buchgr you go, girl! :)

**Comment by normanmaurer** [comment]
@nmittler @buchgr is for sure a strange looking girl, at least from what I remember him looking like ;)

**Comment by buchgr** [comment]
@normanmaurer whaaat? It's the beard isn't it? 😜

**Comment by nmittler** [comment]
@buchgr, @normanmaurer told me that's how you know an Austrian woman is extra "sassy", no? :)

**Comment by buchgr** [comment]
Fixed by c98195714d393a71
